File: sdk/flutter/lib/src/frame.dart
import 'dart:convert';
import 'dart:typed_data';

/// OSP wire protocol constants and frame codec.
class FrameConstants {
  static const magic = <int>[79, 87, 76, 49]; // "OWL1"
  static const protocolVersion = 1;
  static const headerLen = 22;
  static const maxPayloadLen = 16 * 1024 * 1024; // 16 MB

  static const flagCompressed = 1 << 0;
  static const flagChunk = 1 << 1;
  static const flagChunkLast = 1 << 2;
}

/// OSP opcodes.
enum OpCode {
  hello(0x01),
  helloAck(0x02),
  auth(0x03),
  authOk(0x04),
  authFailed(0x05),
  subscribe(0x06),
  subscribeAck(0x07),
  unsubscribe(0x08),
  patch(0x09),
  delete(0x0A),
  restore(0x0B),
  sync(0x0C),
  ack(0x0D),
  heartbeat(0x0E),
  error(0x0F),
  presence(0x10),
  ping(0x11),
  pong(0x12);

  final int value;
  const OpCode(this.value);

  static OpCode? fromValue(int value) {
    for (final op in values) {
      if (op.value == value) return op;
    }
    return null;
  }
}

/// OSP frame header (22 bytes).
class FrameHeader {
  final OpCode opcode;
  final int version;
  final int flags;
  final int length;
  final int reqId;

  FrameHeader({
    required this.opcode,
    this.version = FrameConstants.protocolVersion,
    this.flags = 0,
    required this.length,
    this.reqId = 0,
  });

  /// Encode header to 22 bytes, big-endian.
  Uint8List encode() {
    final data = ByteData(FrameConstants.headerLen);
    int offset = 0;

    // Magic (4 bytes)
    for (final b in FrameConstants.magic) {
      data.setUint8(offset++, b);
    }

    // Version (2 bytes)
    data.setUint16(offset, version, Endian.big);
    offset += 2;

    // Opcode (2 bytes)
    data.setUint16(offset, opcode.value, Endian.big);
    offset += 2;

    // Flags (2 bytes)
    data.setUint16(offset, flags, Endian.big);
    offset += 2;

    // Length (4 bytes)
    data.setUint32(offset, length, Endian.big);
    offset += 4;

    // Request ID (8 bytes)
    data.setInt64(offset, reqId, Endian.big);

    return data.buffer.asUint8List();
  }

  /// Decode header from bytes. Returns null if not enough data.
  static FrameHeader? decode(Uint8List buffer) {
    if (buffer.length < FrameConstants.headerLen) return null;

    final data = ByteData.sublistView(buffer);
    int offset = 0;

    // Check magic
    for (int i = 0; i < 4; i++) {
      if (data.getUint8(offset + i) != FrameConstants.magic[i]) {
        throw FormatException('Invalid magic bytes');
      }
    }
    offset += 4;

    // Version
    final version = data.getUint16(offset, Endian.big);
    offset += 2;
    if (version != FrameConstants.protocolVersion) {
      throw FormatException('Unsupported protocol version: $version');
    }

    // Opcode
    final opcodeRaw = data.getUint16(offset, Endian.big);
    offset += 2;
    final opcode = OpCode.fromValue(opcodeRaw);
    if (opcode == null) {
      throw FormatException('Unknown opcode: 0x${opcodeRaw.toRadixString(16)}');
    }

    // Flags
    final flags = data.getUint16(offset, Endian.big);
    offset += 2;

    // Length
    final length = data.getUint32(offset, Endian.big);
    offset += 4;
    if (length > FrameConstants.maxPayloadLen) {
      throw FormatException('Payload too large: $length');
    }

    // Request ID
    final reqId = data.getInt64(offset, Endian.big);

    return FrameHeader(
      opcode: opcode,
      version: version,
      flags: flags,
      length: length,
      reqId: reqId,
    );
  }
}

/// A complete OSP frame: header + payload.
class Frame {
  final FrameHeader header;
  final Uint8List payload;

  Frame({required this.header, required this.payload});

  Frame.create({
    required OpCode opcode,
    int reqId = 0,
    required this.payload,
    int flags = 0,
  }) : header = FrameHeader(
         opcode: opcode,
         length: payload.length,
         reqId: reqId,
         flags: flags,
       );

  /// Encode entire frame to bytes.
  Uint8List encode() {
    final headerBytes = header.encode();
    final result = Uint8List(headerBytes.length + payload.length);
    result.setAll(0, headerBytes);
    result.setAll(headerBytes.length, payload);
    return result;
  }

  /// Decode frame from bytes. Returns null if not enough data.
  static Frame? decode(Uint8List buffer) {
    final header = FrameHeader.decode(buffer);
    if (header == null) return null;

    final totalLen = FrameConstants.headerLen + header.length;
    if (buffer.length < totalLen) return null;

    final payload = buffer.sublist(FrameConstants.headerLen, totalLen);
    return Frame(header: header, payload: payload);
  }
}

File: sdk/browser/src/frame.ts
/**
 * OSP Frame encoding/decoding for Browser
 * Uses Uint8Array instead of Buffer
 */

export const MAGIC = new Uint8Array([79, 87, 76, 49]); // "OWL1"
export const PROTOCOL_VERSION = 1;
export const HEADER_LEN = 22;
export const MAX_PAYLOAD_LEN = 16 * 1024 * 1024; // 16 MB

export enum OpCode {
  Hello = 0x01,
  HelloAck = 0x02,
  Auth = 0x03,
  AuthOk = 0x04,
  AuthFailed = 0x05,
  Subscribe = 0x06,
  SubscribeAck = 0x07,
  Unsubscribe = 0x08,
  Patch = 0x09,
  Delete = 0x0A,
  Restore = 0x0B,
  Sync = 0x0C,
  Ack = 0x0D,
  Heartbeat = 0x0E,
  Error = 0x0F,
  Presence = 0x10,
  Ping = 0x11,
  Pong = 0x12,
}

export const FLAG_COMPRESSED = 1 << 0;
export const FLAG_CHUNK = 1 << 1;
export const FLAG_CHUNK_LAST = 1 << 2;

export interface FrameHeader {
  opcode: OpCode;
  version: number;
  flags: number;
  length: number;
  reqId: bigint;
}

export interface Frame {
  header: FrameHeader;
  payload: Uint8Array;
}

export function encodeFrame(frame: Frame): Uint8Array {
  const header = new Uint8Array(HEADER_LEN);
  const view = new DataView(header.buffer, header.byteOffset, HEADER_LEN);
  let offset = 0;

  // Magic (4 bytes)
  header.set(MAGIC, offset);
  offset += 4;

  // Version (2 bytes)
  view.setUint16(offset, frame.header.version, false);
  offset += 2;

  // Opcode (2 bytes)
  view.setUint16(offset, frame.header.opcode, false);
  offset += 2;

  // Flags (2 bytes)
  view.setUint16(offset, frame.header.flags, false);
  offset += 2;

  // Length (4 bytes)
  view.setUint32(offset, frame.header.length, false);
  offset += 4;

  // Request ID (8 bytes)
  view.setBigUint64(offset, frame.header.reqId, false);

  // Combine header and payload
  const result = new Uint8Array(HEADER_LEN + frame.payload.length);
  result.set(header, 0);
  result.set(frame.payload, HEADER_LEN);

  return result;
}

export function decodeHeader(buffer: Uint8Array): FrameHeader | null {
  if (buffer.length < HEADER_LEN) {
    return null;
  }

  const view = new DataView(buffer.buffer, buffer.byteOffset, HEADER_LEN);

  // Check magic
  for (let i = 0; i < 4; i++) {
    if (buffer[i] !== MAGIC[i]) {
      throw new Error('Invalid magic bytes');
    }
  }

  let offset = 4;

  const version = view.getUint16(offset, false);
  offset += 2;

  if (version !== PROTOCOL_VERSION) {
    throw new Error(`Unsupported protocol version: ${version}`);
  }

  const opcode = view.getUint16(offset, false) as OpCode;
  offset += 2;

  const flags = view.getUint16(offset, false);
  offset += 2;

  const length = view.getUint32(offset, false);
  offset += 4;

  const reqId = view.getBigUint64(offset, false);

  if (length > MAX_PAYLOAD_LEN) {
    throw new Error(`Payload too large: ${length}`);
  }

  return {
    opcode,
    version,
    flags,
    length,
    reqId,
  };
}

export function decodeFrame(buffer: Uint8Array): Frame | null {
  const header = decodeHeader(buffer);
  if (!header) {
    return null;
  }

  const totalLen = HEADER_LEN + header.length;
  if (buffer.length < totalLen) {
    return null;
  }

  const payload = buffer.slice(HEADER_LEN, totalLen);

  return {
    header,
    payload,
  };
}
